src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,375 @@
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            all_layers.append(new_layer)
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+
+    return (total_swapped, device, all_layers)
+
+
+class XLoraModel(BaseTuner):
+    """
+    Creates an X-LoRA (Mixture of LoRA experts), model from a pretrained transformers model. Currently,
+    this X-LoRA implementation only works with models with a transformer architecture.
+
+    The method is described in detail in https://arxiv.org/abs/2402.07148.
+
+    Args:
+        model ([`torch.nn.Module`]): The model to be adapted.
+        config ([`XLoraConfig`]): The configuration of the Lora model.
+        adapter_name (`str`): The name of the adapter, does not affect the LoRA adapter names.
+
+    Returns:
+        `torch.nn.Module`: The X-LoRA model.
+
+    Example:
+        ```py
+        >>> from transformers import AutoModelForCausalLM, AutoConfig
+        >>> from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_int8_training
+
+        >>> model_config = AutoConfig.from_pretrained("mistralai/Mistral-7B-Instruct-v0.1")
+        >>> config = XLoraConfig(
+        ...     task_type="CAUSAL_LM",
+        ...     hidden_size=model_config.hidden_size,
+        ...     xlora_depth=4,
+        ...     adapters={
+        ...         "adapter_1": "./path/to/the/checkpoint/",
+        ...         "adapter_2": "./path/to/the/checkpoint/",
+        ...         "adapter_n": "./path/to/the/checkpoint/",
+        ...     },
+        ... )
+
+        >>> model = AutoModelForCausalLM.from_pretrained(
+        ...     "mistralai/Mistral-7B-Instruct-v0.1",
+        ...     trust_remote_code=True,
+        ...     use_flash_attention_2=False,
+        ...     device_map="cuda:0",
+        ...     torch_dtype=torch.bfloat16,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> xlora_model = get_peft_model(model, config)
+        ```
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        config: Union[dict[str, XLoraConfig], XLoraConfig],
+        adapter_name: str,
+    ) -> None:
+        nn.Module.__init__(self)
+
+        if isinstance(config, dict):
+            conf = config[adapter_name]
+        else:
+            conf = config
+        lora_model = LoraModel(model, config.copy(), adapter_name)
+        self.xlora_config = conf
+        del self.xlora_config.target_modules
+        self.lora_model = lora_model
+
+        peft_config = conf
+
+        if hasattr(model.config, "use_cache") and model.config.use_cache:
+            raise ValueError("`use_cache` must be False")
+
+        adapters_items = peft_config.adapters.items()
+        if hasattr(self.xlora_config, "_subfolders"):
+            adapters_items = zip(peft_config.adapters.items(), self.xlora_config._subfolders)
```

Comment:
This is never reached in the tests. Is it still relevant or can it be removed?

Reply:
This is the same reasoning as above.